### DISCUSSION TITLE
How to postpone gif animation?

### DISCUSSION THREAD

**Discussion by fAntel** [coil-kt/coil#1419] [Q&A] [answered]
I have `ViewPager` (more precisely `HorizontalPager` from Jetpack Compose) and I have gif for every page. Because `ViewPager` prepare adjacent pages, gif animation begins even when a page is not visible yet. I need to postpone gif animation until a page became current.

My current solution is initialization of `ImageLoader` with `BitmapFactoryDecoder` until gif animation may not be started. But recreation of `ImageLoader` leads to blink before animation begin.
```kotlin
val imageLoader = ImageLoader.Builder(context)
    .components {
        if (!mayBeginGifAnimation) {
            add(BitmapFactoryDecoder.Factory())
        } else if (Build.VERSION.SDK_INT >= 28) {
            add(ImageDecoderDecoder.Factory())
        } else {
            add(GifDecoder.Factory())
        }
    }
    .build()
```

Can I somehow postpone gif animation in Jetpack Compose?

**Comment by colinrtwhite** [comment, marked as answer]
You can call `Animatable.start`/`stop` on the drawable to start and stop the GIF animation. Also I'd strongly avoid creating a new `ImageLoader` for each GIF as it's bad for performance and instead you should share one instance.

**Comment by fAntel** [comment]
Thanks :) It works :) I don't know, there may be a better way to obtain `Animatable` in Jetpack Compose, but I will give here the code that I got, in case someone needs it.
```kotlin
val mayBeginGifAnimation by viewModel.getStates().collectAsState()
var animatable by remember { mutableStateOf<Animatable?>(null) }

animatable?.apply {
    if (!mayBeginGifAnimation) {
        stop()
    } else {
        if (!isRunning) start()
    }
}

AsyncImage(
    model = ImageRequest.Builder(LocalContext.current)
        .repeatCount(0)
        .data(resId)
        .build(),
    modifier = modifier.pointerInput(Unit) {
        // restart gif on tap if it already finished
        detectTapGestures(
            onTap = {
                animatable?.let { animatable ->
                    if (!animatable.isRunning) animatable.start()
                }
            }
        )
    },
    imageLoader = imageLoader,
    onSuccess = { state ->
        (state.result.drawable as? Animatable)?.let {
            animatable = it
        }
    },
    contentDescription = null
)
```
